File: Tools/StoffPy/physicsBox.py
import json
import math
import numpy as np

from geometry_tools import *
from boundary import *
from plane import Plane
import config

class SimulationPoint:
    def __init__(self, pos, vel, acc, fixed):
        self.pos = pos
        self.vel = vel
        self.acc = acc
        self.fixed = fixed

    def __repr__(self):
        return f"({self.pos[0]}, {self.pos[1]})"

class PhysicsBox:
    def __init__(self, fp):
        with open(fp, 'r') as f:
            self.data = json.load(f)

        self.boundaries = []
        self.points = []

        self.target_pt_distance = math.sqrt(config.config()["area_per_pt"]/math.pi) * 2
        vertices = np.array(self.data["vertices"])
        self.vertices = normalize_curve(vertices, self.target_pt_distance)

        self.boundary(self.vertices)
        self.hex()

    def compute_bb(self, vertices):
        min_coords = np.min(vertices, axis=0)
        max_coords = np.max(vertices, axis=0)
        return [min_coords, max_coords]
    
    def boundary(self, vertices):
        if len(self.boundaries) > 0:
            assert self.point_in_boundary(vertices[0]), "Expected new boundary to be contained in the first one"
        else:
            self.bb = self.compute_bb(vertices)

        old_len = len(self.points)
        for i in range(len(vertices)):
            self.points.append(SimulationPoint(
                pos=vertices[i],
                vel=np.array([0.0, 0.0]),
                acc=np.array([0.0, 0.0]),
                fixed=True
            ))

        first_vec = vertices[1] - vertices[0]
        orth = get_orthogonal(first_vec) * 0.00001
        test_pt = vertices[0] + first_vec * 0.5 + orth

        orientation = is_point_in_polygon(test_pt, vertices)

        self.boundaries.append(Boundary(
            box=self,
            start=old_len,
            end=len(self.points),
            orientation=orientation
        ))

    def hex(self):
        assert len(self.boundaries) > 0, "We need a boundary before calling hex"

        triangle_h = math.sqrt(3) / 2 * self.target_pt_distance
        num_lines = int(math.ceil((self.bb[1][1] - self.bb[0][1]) / triangle_h))

        for i in range(num_lines + 1):
            y = self.bb[0][1] + i * triangle_h
            start_x = self.bb[0][0] + (self.target_pt_distance / 2 if i % 2 == 0 else 0)

            x = start_x
            while x <= self.bb[1][0]:
                point = np.array([x, y])
                if self.point_in_boundary(point):
                    self.points.append(SimulationPoint(
                        pos=point,
                        vel=np.array([0.0, 0.0]),
                        acc=np.array([0.0, 0.0]),
                        fixed=False
                    ))
                x += self.target_pt_distance

        return self
    
    def simulation_step(self):
        EPS = 1e-10
        cfg = config.config()
        exp = cfg["gravity_exp"]
        R = math.pow(self.target_pt_distance, exp)  # reference distance

        boundary_points = self.points[self.boundaries[0].start:self.boundaries[0].end]

        for i in range(len(self.points)):
            # Force between points
            for j in range(i + 1, len(self.points)):
                vec = self.points[j].pos - self.points[i].pos
                d = np.linalg.norm(vec)

                if d > cfg["point_force_influence_radius"] * self.target_pt_distance:
                    continue

                acc_amt = R * cfg["point_force_mult"] / math.pow(d + EPS, exp)
                acc = (vec / np.linalg.norm(vec)) * acc_amt

                if np.isnan(acc[0]):
                    acc = np.random.rand(2) * R

                self.points[i].acc -= acc
                self.points[j].acc += acc

            assert not np.isnan(self.points[i].acc[0]), "IsNaN"

            # Force from boundary (only apply if close enough)
            if self.points[i].fixed:
                continue
            for j in range(len(boundary_points)):
                e0 = boundary_points[j].pos
                e1 = boundary_points[(j + 1) % len(boundary_points)].pos

                c = closest_vec_on_line_segment(np.array([e0, e1]), self.points[i].pos)
                d = np.linalg.norm(self.points[i].pos - c)
                if d > self.target_pt_distance * cfg["boundary_force_influence_distance"]:
                    continue

                acc_amt = R * cfg["boundary_force_mult"] / math.pow(d + 0.0000001, exp)
                if np.isnan(acc_amt) or np.isinf(acc_amt):
                    acc_amt = R * 1000

                dir_vec = (e1 - e0) / np.linalg.norm(e1 - e0)
                if not self.boundaries[0].orientation:
                    dir_vec = -dir_vec

                acc = dir_vec * (acc_amt + EPS)

                self.points[i].acc += acc

            assert not np.isnan(self.points[i].acc[0]), "hwa"

        for i in range(len(self.points)):
            if not self.points[i].fixed:
                self.points[i].vel *= cfg["velocity_step_scale"]
                self.points[i].vel += self.points[i].acc

                if np.linalg.norm(self.points[i].vel) > self.target_pt_distance * cfg["velocity_max"]:
                    self.points[i].vel = (self.points[i].vel / np.linalg.norm(self.points[i].vel)) * self.target_pt_distance * cfg["velocity_max"]

                self.points[i].pos += self.points[i].vel
                self.points[i].acc = np.array([0.0, 0.0])

        return self
    
    def point_in_boundary(self, point):
        return is_point_in_polygon(point, self.vertices[self.boundaries[0].start:self.boundaries[0].end])
    
    def filter_expulsed_points(self):
        filtered_points = []
        new_boundaries = []

        currently_in_boundary = False
        boundary_start = 0
        current_index = 0
        boundary_count = 0

        for pt in self.points:
            if currently_in_boundary and not pt.fixed:
                currently_in_boundary = False
                new_boundaries.append(Boundary(
                    box=self,
                    start=boundary_start,
                    end=current_index,
                    orientation=self.boundaries[boundary_count].orientation
                ))
                boundary_count += 1
            elif not currently_in_boundary and pt.fixed:
                currently_in_boundary = True
                boundary_start = current_index

            if pt.fixed or self.point_in_boundary(pt.pos):
                filtered_points.append(pt)
                current_index += 1

        self.points = filtered_points
        self.boundaries = new_boundaries

        return self
    
    def to_plane(self):
        p = Plane()

        p.bb = self.bb  # Copy the bounding box
        p.vertices = [point.pos for point in self.points]  # Copy the vertices

        p.boundaries = self.boundaries  # Copy the boundaries

        return p

File: Tools/StoffGo/shapes2D/physicsBox.go
package shapes2d

import (
	"encoding/json"
	"log"
	"math"
	"math/rand/v2"
	"os"
	"stoffgo/config"
	G "stoffgo/geometry"
	"stoffgo/tools"
)

type ShapeFile struct {
	Vertices []G.Vec2 `json:"vertices"`
}

type SimulationPoint struct {
	pos   G.Vec2
	vel   G.Vec2
	acc   G.Vec2
	fixed bool
}

type PhysicsBox struct {
	Points           []SimulationPoint
	Boundaries       []Boundary
	BB               [2]G.Vec2
	targetPtDistance float64
	// Points to start end end index in points of the fixed boundaries
}

func (p SimulationPoint) Pos() G.Vec2 { return p.pos }
func LoadShapeIntoBox(file string) *PhysicsBox {
	data, err := os.ReadFile(file)
	if err != nil {
		log.Fatalf("Error loading shape file: %v", err)
	}

	var shapeFile ShapeFile
	err = json.Unmarshal(data, &shapeFile)
	if err != nil {
		log.Fatalf("Error loading shape file: %v", err)
	}

	target_pt_distance := math.Sqrt(config.C.Simulation2D.AreaPerPoint/math.Pi) * 2
	shapeFile.Vertices = normalizeCurve(shapeFile.Vertices, target_pt_distance)

	var box PhysicsBox
	box.targetPtDistance = target_pt_distance
	box.boundary(shapeFile.Vertices)
	box.hex()

	return &box
}

func (b *PhysicsBox) FilterExpulsedPoints() *PhysicsBox {
	var filteredPoints []SimulationPoint
	var newBoundaries []Boundary

	currentlyInBoundary := false
	boundaryStart := 0
	currentIndex := 0
	boundaryCount := 0
	for _, pt := range b.Points {
		if currentlyInBoundary && !pt.fixed {
			currentlyInBoundary = false
			newBoundaries = append(newBoundaries, Boundary{boundaryStart, currentIndex, b.Boundaries[boundaryCount].orientation})
			boundaryCount++
		} else if !currentlyInBoundary && pt.fixed {
			currentlyInBoundary = true
			boundaryStart = currentIndex
		}

		if pt.fixed || b.pointInBoundary(pt.pos) {
			filteredPoints = append(filteredPoints, pt)
			currentIndex++
		}
	}

	b.Points = filteredPoints
	b.Boundaries = newBoundaries

	return b
}

func (b *PhysicsBox) boundary(vb []G.Vec2) *PhysicsBox {
	// The first boundary is assumed to be the outer boundary, the rest in the inside
	if len(b.Boundaries) > 0 {
		tools.Assert(
			b.pointInBoundary(vb[0]),
			"Expected new bounday to be contained in first one")
	} else {
		b.BB = computeBB(vb)
	}

	old_len := len(b.Points)
	for i := 0; i < len(vb); i++ {
		b.Points = append(b.Points, SimulationPoint{
			pos:   vb[i],
			vel:   G.Vec2{0, 0},
			acc:   G.Vec2{0, 0},
			fixed: true,
		})
	}

	firstVec := vb[1].Sub(vb[0])
	orth := firstVec.GetOrthogonal().ToLen(0.00001)
	test_pt := vb[0].Add(firstVec.Scale(0.5)).Add(orth)

	b.Boundaries = append(b.Boundaries, Boundary{
		start:       old_len,
		end:         len(b.Points),
		orientation: G.IsPointInPolygon(test_pt, vb),
	})
	return b
}

func (b *PhysicsBox) hex() *PhysicsBox {
	tools.Assert(len(b.Boundaries) > 0, "We need a boundary before calling Hex")

	triangle_h := math.Sqrt(3) / 2 * b.targetPtDistance
	num_lines := int(math.Ceil((b.BB[1][1] - b.BB[0][1]) / triangle_h))

	for i := 0; i < num_lines+1; i++ {
		y := b.BB[0][1] + float64(i)*triangle_h
		var startX float64
		if i%2 == 0 {
			startX = b.BB[0][0] + b.targetPtDistance/2
		} else {
			startX = b.BB[0][0]
		}

		for x := startX; x <= b.BB[1][0]; x += b.targetPtDistance {
			if (b.pointInBoundary(G.Vec2{x, y})) {
				b.Points = append(b.Points, SimulationPoint{
					pos:   G.Vec2{x, y},
					vel:   G.Vec2{0, 0},
					acc:   G.Vec2{0, 0},
					fixed: false,
				})
			}
		}
	}

	return b
}

func (b *PhysicsBox) SimulationStep() *PhysicsBox {
	EPS := 1e-10
	exp := config.C.Simulation2D.GravityExp
	R := math.Pow(b.targetPtDistance, exp) // reference distance

	BoundaryPoints := b.Points[b.Boundaries[0].start:b.Boundaries[0].end]

	for i := range b.Points {
		// Force between points
		for j := i; j < len(b.Points); j++ {
			vec := b.Points[j].pos.Sub(b.Points[i].pos)
			d := vec.Length()

			if d > config.C.Simulation2D.PointForceInfluenceRadius*b.targetPtDistance {
				continue
			}

			accAmt := R * config.C.Simulation2D.PointForceMult / math.Pow(d+EPS, exp)
			acc := vec.ToLen(accAmt)

			if math.IsNaN(acc[0]) {
				acc = G.Vec2{R * rand.Float64(), R * rand.Float64()}
			}

			b.Points[i].acc = b.Points[i].acc.Sub(acc)
			b.Points[j].acc = b.Points[j].acc.Add(acc)
		}

		tools.Assert(!math.IsNaN(b.Points[i].acc[0]), "IsNaN")

		// Force from boundary (only apply if close enough)
		if b.Points[i].fixed {
			continue
		}
		for j := 0; j < len(BoundaryPoints); j++ {
			e0 := BoundaryPoints[j].pos
			e1 := BoundaryPoints[(j+1)%len(BoundaryPoints)].pos

			c := G.ClosestVecOnLineSegment([2]G.Vec2{e0, e1}, b.Points[i].pos)
			d := b.Points[i].pos.Distance(c)
			if d > b.targetPtDistance*config.C.Simulation2D.BoundaryForceInfluenceDistance {
				continue
			}

			accAmt := R * config.C.Simulation2D.BoundaryForceMult / math.Pow(d, exp)
			if math.IsNaN(accAmt) || math.IsInf(accAmt, 1) {
				accAmt = R * 1000
			}

			dir := e1.Sub(e0).GetOrthonormal()
			if !b.Boundaries[0].orientation {
				dir = dir.Scale(-1)
			}

			acc := dir.Scale(accAmt + EPS)

			b.Points[i].acc = b.Points[i].acc.Add(acc)
		}

		tools.Assert(!math.IsNaN(b.Points[i].acc[0]), "hwa")
	}

	for i := range b.Points {
		if !b.Points[i].fixed {
			b.Points[i].vel = b.Points[i].vel.Scale(config.C.Simulation2D.VelocityStepScale)
			b.Points[i].vel = b.Points[i].vel.Add(b.Points[i].acc)

			if b.Points[i].vel.Length() > b.targetPtDistance*config.C.Simulation2D.VelocityMax {
				b.Points[i].vel.ToLen(b.targetPtDistance * config.C.Simulation2D.VelocityMax)
			}

			b.Points[i].pos = b.Points[i].pos.Add(b.Points[i].vel)
			b.Points[i].acc = G.Vec2{0, 0}
		}
	}

	return b
}

func (b *PhysicsBox) ToPlane() *Plane {
	var p Plane
	p.BB = b.BB
	p.Vertices = make([]G.Vec2, len(b.Points))

	for i, point := range b.Points {
		p.Vertices[i] = point.pos
	}

	p.Boundaries = b.Boundaries
	return &p
}
